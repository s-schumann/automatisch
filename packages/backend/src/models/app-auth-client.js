import AES from 'crypto-js/aes.js';
import enc from 'crypto-js/enc-utf8.js';
import appConfig from '../config/app.js';
import Base from './base.js';
import AppConfig from './app-config.js';

class AppAuthClient extends Base {
  static tableName = 'app_auth_clients';

  static jsonSchema = {
    type: 'object',
    required: ['name', 'appKey', 'formattedAuthDefaults'],

    properties: {
      id: { type: 'string', format: 'uuid' },
      appKey: { type: 'string' },
      active: { type: 'boolean' },
      authDefaults: { type: ['string', 'null'] },
      formattedAuthDefaults: { type: 'object' },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static relationMappings = () => ({
    appConfig: {
      relation: Base.BelongsToOneRelation,
      modelClass: AppConfig,
      join: {
        from: 'app_auth_clients.app_key',
        to: 'app_configs.key',
      },
    },
  });

  encryptData() {
    if (!this.eligibleForEncryption()) return;

    this.authDefaults = AES.encrypt(
      JSON.stringify(this.formattedAuthDefaults),
      appConfig.encryptionKey
    ).toString();

    delete this.formattedAuthDefaults;
  }

  decryptData() {
    if (!this.eligibleForDecryption()) return;

    this.formattedAuthDefaults = JSON.parse(
      AES.decrypt(this.authDefaults, appConfig.encryptionKey).toString(enc)
    );
  }

  eligibleForEncryption() {
    return this.formattedAuthDefaults ? true : false;
  }

  eligibleForDecryption() {
    return this.authDefaults ? true : false;
  }

  async triggerAppConfigUpdate() {
    const appConfig = await this.$relatedQuery('appConfig');

    // This is a workaround to update connection allowed column for AppConfig
    await appConfig?.$query().patch({
      key: appConfig.key,
      shared: appConfig.shared,
      disabled: appConfig.disabled,
    });
  }

  // TODO: Make another abstraction like beforeSave instead of using
  // beforeInsert and beforeUpdate separately for the same operation.
  async $beforeInsert(queryContext) {
    await super.$beforeInsert(queryContext);
    this.encryptData();
  }

  async $afterInsert(queryContext) {
    await super.$afterInsert(queryContext);

    await this.triggerAppConfigUpdate();
  }

  async $beforeUpdate(opt, queryContext) {
    await super.$beforeUpdate(opt, queryContext);
    this.encryptData();
  }

  async $afterUpdate(opt, queryContext) {
    await super.$afterUpdate(opt, queryContext);

    await this.triggerAppConfigUpdate();
  }

  async $afterFind() {
    this.decryptData();
  }
}

export default AppAuthClient;
