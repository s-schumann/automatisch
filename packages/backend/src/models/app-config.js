import App from './app.js';
import AppAuthClient from './app-auth-client.js';
import Base from './base.js';

class AppConfig extends Base {
  static tableName = 'app_configs';

  static get idColumn() {
    return 'key';
  }

  static jsonSchema = {
    type: 'object',
    required: ['key'],

    properties: {
      id: { type: 'string', format: 'uuid' },
      key: { type: 'string' },
      connectionAllowed: { type: 'boolean', default: false },
      customConnectionAllowed: { type: 'boolean', default: false },
      shared: { type: 'boolean', default: false },
      disabled: { type: 'boolean', default: false },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static relationMappings = () => ({
    appAuthClients: {
      relation: Base.HasManyRelation,
      modelClass: AppAuthClient,
      join: {
        from: 'app_configs.key',
        to: 'app_auth_clients.app_key',
      },
    },
  });

  async getApp() {
    if (!this.key) return null;

    return await App.findOneByKey(this.key);
  }

  async computeAndAssignConnectionAllowedProperty() {
    this.connectionAllowed = await this.computeConnectionAllowedProperty();
  }

  async computeConnectionAllowedProperty() {
    const appAuthClients = await this.$relatedQuery('appAuthClients');

    const hasSomeActiveAppAuthClients =
      appAuthClients?.some((appAuthClient) => appAuthClient.active) || false;

    const conditions = [
      hasSomeActiveAppAuthClients,
      this.shared,
      !this.disabled,
    ];

    const connectionAllowed = conditions.every(Boolean);

    return connectionAllowed;
  }

  async $beforeInsert(queryContext) {
    await super.$beforeInsert(queryContext);

    await this.computeAndAssignConnectionAllowedProperty();
  }

  async $beforeUpdate(opt, queryContext) {
    await super.$beforeUpdate(opt, queryContext);

    await this.computeAndAssignConnectionAllowedProperty();
  }
}

export default AppConfig;
