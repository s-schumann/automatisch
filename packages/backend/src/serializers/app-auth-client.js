const appAuthClientSerializer = (appAuthClient) => {
  return {
    id: appAuthClient.id,
    appConfigId: appAuthClient.appConfigId,
    name: appAuthClient.name,
    active: appAuthClient.active,
  };
};

export default appAuthClientSerializer;
