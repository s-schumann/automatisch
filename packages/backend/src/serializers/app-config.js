const appConfigSerializer = (appConfig) => {
  return {
    key: appConfig.key,
    customConnectionAllowed: appConfig.customConnectionAllowed,
    shared: appConfig.shared,
    disabled: appConfig.disabled,
    connectionAllowed: appConfig.connectionAllowed,
    createdAt: appConfig.createdAt.getTime(),
    updatedAt: appConfig.updatedAt.getTime(),
  };
};

export default appConfigSerializer;
