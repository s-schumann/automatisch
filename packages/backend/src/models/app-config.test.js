import { vi, describe, it, expect } from 'vitest';

import Base from './base.js';
import AppConfig from './app-config.js';
import App from './app.js';
import AppAuthClient from './app-auth-client.js';
import { createAppConfig } from '../../test/factories/app-config.js';
import { createAppAuthClient } from '../../test/factories/app-auth-client.js';

describe('AppConfig model', () => {
  it('tableName should return correct name', () => {
    expect(AppConfig.tableName).toBe('app_configs');
  });

  it('idColumn should return key field', () => {
    expect(AppConfig.idColumn).toBe('key');
  });

  it('jsonSchema should have correct validations', () => {
    expect(AppConfig.jsonSchema).toMatchSnapshot();
  });

  it('relationMappings should return correct associations', () => {
    const relationMappings = AppConfig.relationMappings();

    const expectedRelations = {
      appAuthClients: {
        relation: Base.HasManyRelation,
        modelClass: AppAuthClient,
        join: {
          from: 'app_configs.key',
          to: 'app_auth_clients.app_key',
        },
      },
    };

    expect(relationMappings).toStrictEqual(expectedRelations);
  });

  describe('getApp', () => {
    it('getApp should return null if there is no key', async () => {
      const appConfig = new AppConfig();
      const app = await appConfig.getApp();

      expect(app).toBeNull();
    });

    it('getApp should return app with provided key', async () => {
      const appConfig = new AppConfig();
      appConfig.key = 'deepl';

      const app = await appConfig.getApp();
      const expectedApp = await App.findOneByKey(appConfig.key);

      expect(app).toStrictEqual(expectedApp);
    });
  });

  describe('computeAndAssignConnectionAllowedProperty', () => {
    it('should call computeConnectionAllowedProperty and assign the result', async () => {
      const appConfig = await createAppConfig();

      const computeConnectionAllowedPropertySpy = vi
        .spyOn(appConfig, 'computeConnectionAllowedProperty')
        .mockResolvedValue(true);

      await appConfig.computeAndAssignConnectionAllowedProperty();

      expect(computeConnectionAllowedPropertySpy).toHaveBeenCalled();
      expect(appConfig.connectionAllowed).toBe(true);
    });
  });

  describe('computeConnectionAllowedProperty', () => {
    it('should return true when app is enabled, shared and allows custom connection with an active app auth client', async () => {
      await createAppAuthClient({
        appKey: 'deepl',
        active: true,
      });

      await createAppAuthClient({
        appKey: 'deepl',
        active: false,
      });

      const appConfig = await createAppConfig({
        disabled: false,
        customConnectionAllowed: true,
        shared: true,
        key: 'deepl',
      });

      const connectionAllowed =
        await appConfig.computeConnectionAllowedProperty();

      expect(connectionAllowed).toBe(true);
    });

    it('should return false if there is no active app auth client', async () => {
      await createAppAuthClient({
        appKey: 'deepl',
        active: false,
      });

      const appConfig = await createAppConfig({
        disabled: false,
        customConnectionAllowed: true,
        shared: true,
        key: 'deepl',
      });

      const connectionAllowed =
        await appConfig.computeConnectionAllowedProperty();

      expect(connectionAllowed).toBe(false);
    });

    it('should return false if there is no app auth clients', async () => {
      const appConfig = await createAppConfig({
        disabled: false,
        customConnectionAllowed: true,
        shared: true,
        key: 'deepl',
      });

      const connectionAllowed =
        await appConfig.computeConnectionAllowedProperty();

      expect(connectionAllowed).toBe(false);
    });

    it('should return false when app is disabled', async () => {
      const appConfig = await createAppConfig({
        disabled: true,
        customConnectionAllowed: true,
      });

      const connectionAllowed =
        await appConfig.computeConnectionAllowedProperty();

      expect(connectionAllowed).toBe(false);
    });

    it(`should return false when app doesn't allow custom connection`, async () => {
      const appConfig = await createAppConfig({
        disabled: false,
        customConnectionAllowed: false,
      });

      const connectionAllowed =
        await appConfig.computeConnectionAllowedProperty();

      expect(connectionAllowed).toBe(false);
    });
  });

  it('$beforeInsert should call computeAndAssignConnectionAllowedProperty', async () => {
    const computeAndAssignConnectionAllowedPropertySpy = vi
      .spyOn(AppConfig.prototype, 'computeAndAssignConnectionAllowedProperty')
      .mockResolvedValue(true);

    await createAppConfig();

    expect(computeAndAssignConnectionAllowedPropertySpy).toHaveBeenCalledOnce();
  });

  it('$beforeUpdate should call computeAndAssignConnectionAllowedProperty', async () => {
    const appConfig = await createAppConfig();

    const computeAndAssignConnectionAllowedPropertySpy = vi
      .spyOn(AppConfig.prototype, 'computeAndAssignConnectionAllowedProperty')
      .mockResolvedValue(true);

    await appConfig.$query().patch({
      key: 'deepl',
    });

    expect(computeAndAssignConnectionAllowedPropertySpy).toHaveBeenCalledOnce();
  });
});
