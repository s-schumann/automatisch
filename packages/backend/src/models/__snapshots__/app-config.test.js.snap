// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppConfig model > jsonSchema should have correct validations 1`] = `
{
  "properties": {
    "connectionAllowed": {
      "default": false,
      "type": "boolean",
    },
    "createdAt": {
      "type": "string",
    },
    "customConnectionAllowed": {
      "default": false,
      "type": "boolean",
    },
    "disabled": {
      "default": false,
      "type": "boolean",
    },
    "id": {
      "format": "uuid",
      "type": "string",
    },
    "key": {
      "type": "string",
    },
    "shared": {
      "default": false,
      "type": "boolean",
    },
    "updatedAt": {
      "type": "string",
    },
  },
  "required": [
    "key",
  ],
  "type": "object",
}
`;
